// cmd-transport/lib/cmd-transport.js
/**
 * @description
 * 由于是依赖gruntjs构建工具，在做transport前，需要创建Gruntfile.js
 */

var FS = require('fs');
var PATH = require('path');


module.exports = createGruntfile;


/**
 * @param {string} src
 * @param {string} dest
 * @param {object} alias
 * @param {function} callback
 * @returns {undefined}
 */
function createGruntfile(src, dest, alias, gruntPath, callback) {

    var transport = {
        proj: {
            options: {
                debug: false,
                alias: alias
            },
            files: [
                {
                    expand: true,
                    cwd: src,
                    src: "**/*.js",
                    dest: dest
                }
            ]
        }
    };

    var output = '\
        module.exports = function(grunt){\
            grunt.initConfig({\
                pkg: grunt.file.readJSON("package.json"),\
                transport:' + JSON.stringify(transport) +
            '});\
            grunt.loadNpmTasks("grunt-cmd-transport");\
            grunt.registerTask("default", ["transport"]);\
        };';
    output = output.replace(/\s{4}/g, '');


    FS.writeFileSync(PATH.join(gruntPath,'Gruntfile.js'), output, 'utf-8');
    execGrunt(gruntPath, callback);

}



/**
 * @param {string} path gruntjs所在的路径
 * @param {function} callback
 * @returns {}
 */
function execGrunt(path, callback) {
    var exec = require('child_process').exec;
    exec('cd ' + path + ' && grunt --force', function(error, stdout, stderr) {
        if (error !== null) {
            console.log('exec error: ' + error);
            return;
        }
        console.log(stdout);
        typeof callback === 'function' ? callback() : null;
    });
}
