// cmd-transport/lib/cmd-transport.js
/**
 * @description
 * 由于是依赖gruntjs构建工具，在做transport前，需要创建Gruntfile.js
 */

var FS = require('fs');
var PATH = require('path');
var mv = require('util-mv');


module.exports = transportTask;


/**
 * @param {string} src
 * @param {string} dest
 * @param {object} alias
 * @param {function} callback
 * @returns {undefined}
 */
function transportCopy(src, dest){
	var ll = FS.readdirSync(src);
	ll.forEach(function(file){
		var filename = PATH.join(src, file);
		var destFilename = PATH.join(dest, file);
		if(FS.statSync(filename).isDirectory()){
			transportCopy(filename, destFilename)
		}else{
			mv(filename, destFilename);
		}
	});
}


/**
 * @param {string} src
 * @param {string} dest
 * @param {object} alias
 * @param {function} callback
 * @returns {undefined}
 */
function transportTask(src, dest, alias, gruntPath, callback) {

	// 先复制源文件
	transportCopy(src, dest);
	
	var transport = "{\
		transport4js: {\
            options: {\
                debug: false,\
                alias: " + JSON.stringify(alias) + "\
            },\
            files: [\
                {\
                    expand: true,\
                    cwd: '" + src + "',\
                    src: '**/*.js',\
                    dest: '" + dest + "'\
                }\
            ]\
        },\
		transport4css: {\
			options: {\
				debug: false,\
				parsers: {\
					'.css': [css2jsParser]\
				}\
			},\
			files: [{\
				expand: true,\
				cwd: '" + src + "',\
				src: '**/*.css',\
				dest: '" + dest + "'\
			}]\
		}\
	}";
	
    var output = '\
        module.exports = function(grunt){\
			var style = require("grunt-cmd-transport").style.init(grunt);\
			var css2jsParser = style.css2jsParser;\
			var jsParser = require("grunt-cmd-transport").script.init(grunt).jsParser;\
            grunt.initConfig({\
                pkg: grunt.file.readJSON("package.json"),\
                transport:' + transport + 
            '});\
            grunt.loadNpmTasks("grunt-cmd-transport");\
            grunt.registerTask("default", ["transport"]);\
        };';
    output = output.replace(/\s{4}/g, '');

    FS.writeFileSync(PATH.join(gruntPath,'Gruntfile.js'), output, 'utf-8');
    execGrunt(gruntPath, callback);
}



/**
 * @param {string} path gruntjs所在的路径
 * @param {function} callback
 * @returns {}
 */
function execGrunt(path, callback) {
    var exec = require('child_process').exec;
    exec('cd ' + path + ' && grunt --force', function(error, stdout, stderr) {
        if (error !== null) {
            console.log('exec error: ' + error);
            return;
        }
        // console.log(stdout);
        typeof callback === 'function' ? callback() : null;
    });
}
