// build-task/test/proj/resource/page/app/restaurant/js/list.js

define(function (require, exports, module) {

	var $ = require('jquery');

	require('ipick.data-table');
	require('ipick.advanced-pagination');

	$('#restaurant-table').dataTable();
	$('#table-pagination').advancedPagination();

	// Search
	$('#search-form').submit(function (event) {
		event.preventDefault();
		$('#current-keyword').val($('#keyword').val());
		$('#table-pagination').trigger('pagination');
	});

	// Pagination
	$('#table-pagination').on('pagination', function (event, callback) {
		$.getJSON('search', {
			'keyword': $('#current-keyword').val(),
			'page': event.page
		}, function (data) {
			var baseIndex = data['data']['data']['page_info']['num'] *
				(data['data']['data']['page_info']['page'] - 1) + 1;
			var rows = $.map(data['data']['data']['list'], function (item, index) {
				var row = {
					id: [
						{
							type: 'text',
							text: baseIndex + index
						},
						{
							type: 'hidden',
							clazz: 'item-id',
							value: item['_id_h']
						}
					],
					telephone: item['tel'],
					name: item['name'],
					address: item['address'],
					edit: [
						{
							type: 'link',
							clazz: 'modify-btn',
							href: 'modify?id=' + item['id'],
							text: '<span class="glyphicon glyphicon-cog"></span>',
							tooltip: 'Modify'
						},
						{
							type: 'text',
							text: '&nbsp;'
						},
						{
							type: 'link',
							clazz: 'remove-btn',
							href: 'javascript:void(0)',
							text: '<span class="glyphicon glyphicon-trash"></span>',
							tooltip: 'Remove'
						}
					]
				};
				return row;
			});
			$('#restaurant-table')
				.dataTable().clear()
				.dataTable().addRow(rows);
			$('#table-pagination')
				.advancedPagination().reload({
					url: 'javascript:void(0)',
					min: 1,
					max: data['data']['data']['page_info']['total_page'],
					range: 3,
					current: data['data']['data']['page_info']['page'],
				});
			if (callback) callback();
		});
	});

});
