// build-task/test/proj/resource/page/app/restaurant/js/verify.modify.js
/**
 * @author
 * @desc
 */
define(function(require, exports, module){
	var $ = require('jquery');
	var IVYAlert = require('ivy.alert');
	require('validation');
	require('bootstrap'); 
    require('select2');
	
    // 初始化需要多语言添加的表单项
    // var IPICKFclangname = require('ipick.fc.langname');
    // var category = new IPICKFclangname();
    // category.init('#category', $('#category').children().children(),{
    //     add_btn_txt: 'Add Category',
    //     data: $('#category').attr('data-val') ? JSON.parse($('#category').attr('data-val')) : []
    // });
	
	// var SWFUpload = require('swfupload');
	// var swfConfig = require('page/app/search/js/search.modify.swf.config');

	// // Icon
	// new SWFUpload({
	// 	// Backend Settings
	// 	upload_url: "/cms/index.php/upload/up2tfs",
	// 	post_params: {
	// 		'user_name': 's_id'
	// 	},

	// 	// File Upload Settings
	// 	file_size_limit : "2 MB",	// 2MB
	// 	file_types : "*.jpg;*.png;*.gif",
	// 	file_types_description : "JPG/PNG/GIF Images",
	// 	file_upload_limit : "0",

	// 	file_queue_error_handler : swfConfig.file_queue_error_handler,
	// 	file_dialog_complete_handler : swfConfig.file_dialog_complete_handler,
	// 	upload_progress_handler : swfConfig.upload_progress_handler, //
	// 	upload_error_handler : swfConfig.upload_error_handler,
	// 	upload_success_handler : function(file, rs){
	// 		try{
	// 			rs = JSON.parse(rs);
	// 			if(rs.ret == 0){
	// 				$('#category_icon').val(rs.data.pic_url);
	// 				$('#upload_img').removeClass('hidden').attr('src', rs.data.pic_url);
	// 				//网络慢的情况下，提示信息后很久才看到图片
	// 				IVYAlert('Upload successfully!');
	// 			}else{
	// 				IVYAlert(rs.msg);
	// 			}
	// 		}catch(e){
	// 			IVYAlert(rs);
	// 		}
	// 	},
	// 	upload_complete_handler : swfConfig.upload_complete_handler,

	// 	button_placeholder_id : "upload",
	// 	button_width: 1000,
	// 	button_height: 300,
	// 	button_window_mode: SWFUpload.WINDOW_MODE.TRANSPARENT,
	// 	button_cursor: SWFUpload.CURSOR.HAND,

	// 	// Flash Settings
	// 	flash_url : "/cms/resource/lib/cmd/swf/swfupload.swf",

	// 	// Debug Settings
	// 	debug: false
	// });



    var restaurant = {
        edit: function(){
            this.get2Cuisine();
            this.get2Region();
            this.getLandmarkList();
            this.addDiscount();
        },

        list: function(){
            this.bindListAction();

        },
 
        /**
         * 二级菜式请求
         */
        get2Cuisine: function(){
            $('#subTypeF').change(function(){
                $('#subTypeS').attr('disabled', '');
                $.getJSON('/cms/index.php/cuisine/get_cuisine', {
                    id: this.value
                }, function(rs){
                    if(rs.ret == 0){
                        rs.data.data.cuisine_list ? render2DishesOptions(rs.data.data.cuisine_list) : null;
                    }else{
                        IVYAlert(rs.msg);
                    }
                })
            }).change();

            function render2DishesOptions(data){
                var options = "";
                for(var i = 0,l = data.length;i < l;i++){
                    options += '<option value="' + data[i].id + '" data-name="' + data[i].name + '">' + data[i].name + '</option>'
                }
                var $target = $('#subTypeS');
                $target.html(options);
                //$target.removeAttr('disabled');
                if($target.attr('data-val')){
                    $target.children('[data-name="' + $target.attr('data-val') + '"]').attr('selected', '');
                    $target.removeAttr('data-val');
                }
            }
        },
       

        /**
         * 二级地区请求
         */
        get2Region: function(){
            $('#regionF').change(function(){
                $('#regionS').attr('disabled', '');
                $.getJSON('/cms/index.php/region/get_region', {
                    id: this.value
                }, function(rs){
                    if(rs.ret == 0){
                        rs.data.data.region_list ? render2RegionOptions(rs.data.data.region_list) : null;
                    }else{
                        IVYAlert(rs.msg);
                    }
                })
            }).change();

            function render2RegionOptions(data){
                var options = "";
                for(var i = 0,l = data.length;i < l;i++){
                    options += '<option value="' + data[i].id + '" data-name="' + data[i].name + '">' + data[i].name + '</option>'
                }
                var $target = $('#regionS');
                $target.html(options);
                //$target.removeAttr('disabled');
                if($target.attr('data-val')){
                    $target.children('[data-name="' + $target.attr('data-val') + '"]').attr('selected', '');
                    $target.removeAttr('data-val');
                }
            }
        },


        /**
         * 添加折扣优惠
         */
        addDiscount: function(){
            $('#discountFL').change(function(){
                if(this.value === ""){
                    return;
                }

                $('#discountSL, #discountTL').children().remove();
                $.getJSON('/cms/index.php/discount/get_bank_type_list', {
                    bank_name: $(this).val()
                }, function(rs){
                    if(rs.ret === 0){
                        $('#discountSL, #discountTL').children().remove();
                        $('#discountSL').html(renderSLSelectData(rs.data.card_list));
                    }else{
                        IVYAlert(rs.msg);
                    }
                });
            });

            $('#discountSL').change(function(){
                if(this.value === ""){
                    return;
                }

                $.getJSON('/cms/index.php/discount/get_discount_list', {
                    bank_name: $('#discountFL').val(),
                    card_type: $(this).val()
                }, function(rs){
                    if(rs.ret === 0){
                        $('#discountTL').children().remove();
                        $('#discountTL').html(renderTLSelectData(rs.data.discount_list));
                    }else{
                        IVYAlert(rs.msg);
                    }
                });
            });

            function renderSLSelectData(data){
                if(!data){
                    return '';
                }
                var tpl = '<option value="{v}">{v}</option>'
                var output = '';
                for(var i = 0,l = data.length;i < l;i++){
                    output += tpl.replace(/{v}/g, data[i]);
                }
                return '<option value="">---</option>' + output;
            }

            function renderTLSelectData(data){
                if(!data){
                    return '';
                }
                var tpl = '<option value="{val}">{txt}</option>'
                var output = '';
                for(var itm in data){
                    output += tpl.replace(/{val}/, itm).replace(/{txt}/, data[itm]);
                }
                return output;
            }


            // 添加折扣
            $('#addDiscount').click(function(){
                var tpl = '<div class="alert alert-success alert-dismissible" role="alert" style="margin-bottom:5px;padding: 5px 35px 5px 15px;">\
                                <button type="button" class="close" data-dismiss="alert">\
                                    <input type="hidden" value="{val}"/>\
                                    <span aria-hidden="true">&times;</span>\
                                    <span class="sr-only" data-rel="discount-remove">Close</span>\
                                </button>\
                                <small><span class="glyphicon glyphicon-tags"></span> {bank}<strong>|</strong>{card}<strong>|</strong>{discount}</small>\
                            </div>';
                $(tpl.replace('{bank}', $('#discountFL').val())
                    .replace('{card}', $('#discountSL').val())
                    .replace('{discount}', $('#discountTL').children('[value="' + $('#discountTL').val() + '"]').text())
                    .replace('{val}', $('#discountTL').val()))
                    .appendTo($('#discountList'));
            });
            // 监听删除折扣
            $('#discountList').delegate('a[data-rel="discount-remove"]', 'click', function(){
                $(this).parent().remove();
            })
        },

        getDiscountData: function(){
            var output = [];
            $('#discountList').children().each(function(){
                output.push($(this).find('input').val());
            });
            return output.length === 0 ? "" : JSON.stringify(output);
        },
   

        /**
         * 地标选择
         */
        getLandmarkList: function(){

            $('#landmark').select2({
                multiple: true,
                separator: '##',
                minimumInputLength: 1,
                formatInputTooShort: 'Please enter the landmark',
                initSelection: function (element, callback) {
                    element.val('');
                    callback(encodeTagsData(JSON.parse($('#landmarkids').val())));
                },
                ajax: {
                    url: '/cms/index.php/landmark/search',
                    data: function (term) {
                        return { keyword: term };
                    },
                    results: function (data) {
                        return {
                            results: encodeTagsData(data.data.data.all_list)
                        };
                    }
                }
            }).on('change', function (event) {
                $('#landmarkids').val(JSON.stringify(decodeTagsData($('#landmark').val()).get()));
            });               




        },

        /**
         * 获取地标数据
         */
        getLandmarkData: function(){
            var rs = [];
            var obj = JSON.parse($('#landmarkids').val());

            for(var i=0; i < obj.length; i++ ){
                rs.push(obj[i].id);
            }
            return rs;
        },

        bindListAction: function(){
            $('.btn_commit').click(function(e){
                var id = $(this).attr('data-val');
                $.post('/cms/index.php/restaurant/push_rest', {
                    id: id
                }, function(rs){

                    if(rs.ret){
                        //TODO;
                    }
                },'json')
            })
        }

    }


    $('#restaurant-register-form').validate({
        rules: {
            // desc: 'required'
        },
        errorElement: 'span',
        errorPlacement: function(error,el){
        },
        submitHandler:function(form) {
            IVYAlert("Loading...", {
                toggle: true
            });
            var data = {
                id: form.id.value,
                type: $('#status > label.active > input').val(),
                weight: $('#weight').val(),
                editorial_weight: $('#editorial_weight').val(),
                description: form.desc.value
            };

            if(!data.type){
                IVYAlert('Need review status');
                return;
            }

            $.post('/cms/index.php/rest_verify/approval', data, function(rs){
                if(rs.ret == 0){
                    IVYAlert(rs.msg, function(){
                        location.href = "/cms/index.php/rest_verify/index";
                    });
                }else{
                    IVYAlert(rs.msg);
                }
            },'json');
        }
    });

	
    function getI18nJSON(selector){
        var obj = {};
        $(selector).parent().find('input, textarea').each(function(i, e){     
            var key = $(e).data('lang');
            var val = $(e).val();
            obj[key] = val;
        })
        return JSON.stringify(obj);
    }

    function encodeTagsData(data) {
        return $(data || []).map(function (index, tag) {
            return {
                id: JSON.stringify(
                    {id: tag['id'], name: tag['name']}),
                text: tag['name']
            };
        });
    };
    function decodeTagsData(string) {
        return $(string.split('##')).map(function (index, tagString) {
            return tagString ? JSON.parse(tagString) : '[]';
        });
    };
	
	restaurant.edit();
})
