// build-task/test/proj/resource/page/app/restaurant/js/verify.index.js
/**
 * @author
 * @desc
 */
define(function(require, exports, module){
    var $ = require('jquery');
    var IVYAlert = require('ivy.alert');
    var Pagingbar = require('pagingbar');

    // Rewrite goto
    Pagingbar.gotoPage = function(pnum){
        if(pnum > 0 && pnum <= Pagingbar._data.total_page_num){
            Pagingbar.setPageNum(pnum);
            // Set index ui
            $('#common-pagingbar').find('li').removeClass('active');
            $('#common-pagingbar').find('a[data-val="' + pnum + '"]').parent().addClass('active');
            $.get('/cms/index.php/rest_verify/get_more', {
                page: Pagingbar.getPageNum(),
                num : Pagingbar.getPageSize()
            }, function(rs){
                $(window).scrollTop(0);
                var list = renderList(rs.data.data.list);
                if(list !== ""){
                    $('#homeList').children().remove();
                    $('#homeList').html(list);
                }
            },'json')
        }else{
            $('#common-pagingbar-goto').focus();
        }
    }
    // Init paging
    Pagingbar.init({
        page_size:  parseInt($('#common-pagingbar-page_size').val()),
        page_num:   parseInt($('#common-pagingbar-page_num').val()),
        total_page_num: parseInt($('#common-pagingbar-total_page_num').val()),
        country: $('#country').val()
    });
    /**
     * 刷新列表
     */
    var tplListData = require('page/app/restaurant/js/verify.index.data.js');
    function renderList(data){
        if(!data || data.length === 0){
            IVYAlert('No data');
            return "";
        }else{
            var status = {"1":"本地保存", "2":"提交审核", "3":"审核成功", "4":"审核失败"};
            var html = "";
            var _idx = (Pagingbar.getPageNum() - 1) * Pagingbar.getPageSize() + 1;
            for(var i = 0, l = data.length;i < l;i++){
                html += tplListData.replace(/{index}/g, _idx + i)
                .replace(/{name}/g, data[i].name)
                .replace(/{address}/g, data[i].address)
                .replace(/{last_update_h}/g, data[i].last_update_h)
                .replace(/{status}/g, status[data[i].status])
                .replace(/{id}/g, data[i]._id_h);
            }
            return html;
        }

        function disposeJSON(data){
            var rs = "";
            for(var itm in data){
                rs += '<div>' + itm + ': ' + (data[itm] ? data[itm] : '') + '</div>';
            }
            return rs;
        }
    }


})
