// build-task/test/proj/resource/page/app/distance/distance.modify.js
/**
 * @author
 * @desc
 */
define(function(require, exports, module){
    var $ = require('jquery');
    var IVYAlert = require('ivy.alert');
    require('validation');
    var IPICKFclangname = require('ipick.fc.langname');

    // 初始化需要多语言添加的表单项
    var category = new IPICKFclangname();
    category.init('#category', $('#category').children().children(),{
        add_btn_txt: 'Add Category',
        data: $('#category').attr('data-val') ? JSON.parse($('#category').attr('data-val')) : []
    });

    var minDistance = new IPICKFclangname();
    minDistance.init('#minDistance', $('#minDistance').children().children(),{
        add_btn_txt: 'Add Item',
        data: $('#minDistance').attr('data-val') ? JSON.parse($('#minDistance').attr('data-val')) : []
    });

    var maxDistance = new IPICKFclangname();
    maxDistance.init('#maxDistance', $('#maxDistance').children().children(),{
        add_btn_txt: 'Add Item',
        data: $('#maxDistance').attr('data-val') ? JSON.parse($('#maxDistance').attr('data-val')) : []
    });

    $('#form').validate({
        rules: {
            order: {
                required: true
            },
            main_order: {
                required: true
            }
        },
        errorElement: 'span',
        errorPlacement: function(error,el){
            el.parent().find('[data-control="form-input-error"]').append(error);
        },
        submitHandler:function(form) {
            IVYAlert("Loading...", {
                toggle: true
            });

            var data = {
                id: form.id.value,
                order: form.order.value,
                main_order: form.main_order.value,
                category_value: category.getData(),
                distance_min: minDistance.getData(),
                distance_max: maxDistance.getData()
            };

            if(data.max_distance === ""){
                IVYAlert('Max Distance is required!');
                return;
            }

            $.post('/cms/index.php/distance/do_modify', data, function(rs){
                if(rs.ret == 0){
                    IVYAlert(rs.msg, function(){
                        location.href = "/cms/index.php/distance/index";
                    });
                }else{
                    IVYAlert(rs.msg);
                }
            },'json');
        }
    });
})
