// build-task/test/proj/resource/page/app/blogger/js/blogger.list.js
/**
 * @author
 * @desc
 */
define(function(require, exports, module){
    var $ = require('jquery');
    var IVYAlert = require('ivy.alert');
    var Pagingbar = require('pagingbar');

    // Rewrite goto
    Pagingbar.gotoPage = function(pnum){
        if(pnum > 0 && pnum <= Pagingbar._data.total_page_num){
            Pagingbar.setPageNum(pnum);
            // Set index ui
            $('#common-pagingbar').find('li').removeClass('active');
            $('#common-pagingbar').find('a[data-val="' + pnum + '"]').parent().addClass('active');
            $.get('/cms/index.php/article/get_more', {
                page: Pagingbar.getPageNum(),
                num : Pagingbar.getPageSize()
            }, function(rs){
                $(window).scrollTop(0);
                var list = renderList(rs.data.data.list);
                if(list !== ""){
                    $('#homeList').children().remove();
                    $('#homeList').html(list);
                }
            },'json')
        }else{
            $('#common-pagingbar-goto').focus();
        }
    }
    // Init paging
    Pagingbar.init({
        page_size:  parseInt($('#common-pagingbar-page_size').val()),
        page_num:   parseInt($('#common-pagingbar-page_num').val()),
        total_page_num: parseInt($('#common-pagingbar-total_page_num').val()),
        country: $('#country').val()
    });
    /**
     * 刷新列表
     */
    var tplListData = require('page/app/blogger/js/blogger.list.data.js');
    var status = ["", "", "正在审核", "审核成功", "审核失败"];
    function renderList(data){
        if(!data || data.length === 0){
            IVYAlert('No data');
            return "";
        }else{
            var html = "";
            var _idx = (Pagingbar.getPageNum() - 1) * Pagingbar.getPageSize() + 1;
            for(var i = 0, l = data.length;i < l;i++){
                html += tplListData.replace(/{index}/g, _idx + i)
                    .replace(/{cgi_id}/g, data[i].cgi_id)
                    .replace(/{blogger_name}/g, data[i].blogger_info && data[i].blogger_info.hasOwnProperty('blogger_name')?data[i].blogger_info.blogger_name:'')
                    .replace(/{title}/g, data[i].title)
                    .replace(/{rest}/g, data[i].rest)
                    .replace(/{last_update_h}/g, data[i].last_update_h)
                    .replace(/{status}/g, status[data[i].status])
                    .replace(/{id}/g, data[i]._id_h)
            }
            return html;
        }
    }

})
