// build-task/test/proj/resource/page/app/blogger/js/blogger.modify.js
/**
 * @author
 * @desc
 */
define(function(require, exports, module){
    var $ = require('jquery');
    var IVYAlert = require('ivy.alert');
    require('validation');
    require('bootstrap');

    $('#form').validate({
        rules: {
            desc: 'required'
        },
        errorElement: 'span',
        errorPlacement: function(error,el){
        },
        submitHandler:function(form) {
            IVYAlert("Loading...", {
                toggle: true
            });
            var data = {
                id: form.id.value,
                type: $('#status > label.active > input').val(),
                desc: form.desc.value
            };

            if(!data.type){
                IVYAlert('Need review status');
                return;
            }

            $.post('/cms/index.php/article/approval', data, function(rs){
                if(rs.ret == 0){
                    IVYAlert(rs.msg, function(){
                        location.href = "/cms/index.php/article/index";
                    });
                }else{
                    IVYAlert(rs.msg);
                }
            },'json');
        }
    });
})
