// build-task/test/proj/resource/page/app/home/js/index.js

define(function (require, exports, module) {

	var $ = require('jquery');
	var IVYAlert = require('ivy.alert');

	require('ipick.home-preview');
	require('ipick.data-table');
	require('ipick.advanced-pagination');

	$('#modified-preview, #live-preview').homePreview();
	$('#rule-table').dataTable();
	$('#table-pagination').advancedPagination();

	// Pagination
	$('#table-pagination').on('pagination', function (event, callback) {
		$.getJSON('get_more', {
			'page': event.page
		}, function (data) {
			var baseIndex = data['data']['data']['page_info']['num'] *
				(data['data']['data']['page_info']['page'] - 1) + 1;
			var rows = $.map(data['data']['data']['list'], function (item, index) {
				var row = {
					id: [
						{
							type: 'text',
							text: baseIndex + index
						},
						{
							type: 'hidden',
							clazz: 'item-id',
							value: item['_id_h']
						}
					],
					title: item['display_name'],
					icon: [
						{
							type: 'image',
							src: item['icon'],
							alt: item['display_name']
						}
					],
					type: {
						'1': 'Tags',
						'2': 'Manually',
						'3': 'Articles'
					}[item['rule_type']],
					status: [
						{
							type: 'link',
							clazz: 'text-danger enable-btn ' + (item['status'] != 0 ? 'hidden' : ''),
							href: 'javascript:void(0)',
							text: '<span class="glyphicon glyphicon-remove"></span>',
							tooltip: 'Click to Enable'
						},
						{
							type: 'link',
							clazz: 'text-success disable-btn ' + (item['status'] == 0 ? 'hidden' : ''),
							href: 'javascript:void(0)',
							text: '<span class="glyphicon glyphicon-ok"></span>',
							tooltip: 'Click to Disable'
						},
						{
							type: 'hidden',
							clazz: 'item-status',
							value: item['status']
						}
					],
					lastModified: item['last_update'],
					edit: [
						{
							type: 'link',
							clazz: 'modify-btn',
							href: 'modify?id=' + item['_id_h'],
							text: '<span class="glyphicon glyphicon-cog"></span>',
							tooltip: 'Modify'
						},
						{
							type: 'text',
							text: '&nbsp;'
						},
						{
							type: 'link',
							clazz: 'remove-btn',
							href: 'javascript:void(0)',
							text: '<span class="glyphicon glyphicon-trash"></span>',
							tooltip: 'Remove'
						}
					],
					weight: [
						{
							type: 'textbox',
							clazz: 'form-control input-sm text-center weight-input',
							value: item['show_index'],
							width: 50
						}
					]
				};
				return row;
			});
			$('#rule-table')
				.dataTable().clear()
				.dataTable().addRow(rows);
			if (callback) callback();
		});
	});

	// Enable/Disable
	$('#rule-table').delegate('.enable-btn', 'click', function (event) {
		var tableRow = $(this).parents('tr').first();
		$.getJSON('online', {
			'id': tableRow.find('.item-id').val()
		}, function (data) {
			if (data['ret']) {
				IVYAlert(data['msg']);
				// $('#error-message').html('Sorry, your submission failed. Please try again.');
				// $('#error-message').removeClass('hidden');
			} else {
				tableRow.find('.enable-btn').addClass('hidden');
				tableRow.find('.disable-btn').removeClass('hidden');
				var items = $.map(data['data']['online_list'], function (item, index) {
					return {
						title: item['display_name'],
						iconUrl: item['icon']
					};
				});
				$('#modified-preview')
					.homePreview().clear()
					.homePreview().addItem(items);
			}
		});
	});
	$('#rule-table').delegate('.disable-btn', 'click', function (event) {
		var tableRow = $(this).parents('tr').first();
		$.getJSON('offline', {
			'id': tableRow.find('.item-id').val()
		}, function (data) {
			if (data['ret']) {
				IVYAlert(data['msg']);
				// $('#error-message').html('Sorry, your submission failed. Please try again.');
				// $('#error-message').removeClass('hidden');
			} else {
				tableRow.find('.disable-btn').addClass('hidden');
				tableRow.find('.enable-btn').removeClass('hidden');
				var items = $.map(data['data']['online_list'], function (item, index) {
					return {
						title: item['display_name'],
						iconUrl: item['icon']
					};
				});
				$('#modified-preview')
					.homePreview().clear()
					.homePreview().addItem(items);
			}
		});
	});

	// Sort
	$('#rule-table').delegate('.weight-input', 'change', function (event) {
		var tableRow = $(this).parents('tr').first();
		var weight = parseInt($(this).val());
		if (isNaN(weight)) return false;
		$.getJSON('modify_index', {
			'id': tableRow.find('.item-id').val(),
			'show_index': weight
		}, function (data) {
			if (data['ret']) {
				IVYAlert(data['msg']);
				// $('#error-message').html('Sorry, your submission failed. Please try again.');
				// $('#error-message').removeClass('hidden');
			} else {
				var items = $.map(data['data']['online_list'], function (item, index) {
					return {
						title: item['display_name'],
						iconUrl: item['icon']
					};
				});
				$('#modified-preview')
					.homePreview().clear()
					.homePreview().addItem(items);
			}
		});
	});

	// Release
	$('#release-btn').click(function (event) {
		$.getJSON('../release/home_search',
		function (data) {
			if (data['ret']) {
				IVYAlert(data['msg']);
				// $('#error-message').html('Sorry, your submission failed. Please try again.');
				// $('#error-message').removeClass('hidden');
			} else {
				IVYAlert(data['msg']);
				// $('#success-message').html('Your edit has been submitted.');
				// $('#success-message').removeClass('hidden');
				$('#live-preview')
					.homePreview().clear()
					.homePreview().addItem($('#modified-preview')
						.homePreview().getItemList()
					);
			}
		});
	});

});
