// build-task/test/proj/resource/page/app/discount/discount.modify.js
/**
 * @author
 * @desc
 */
define(function(require, exports, module){
    var $ = require('jquery');
    var IVYAlert = require('ivy.alert');
    require('validation');

    $('#form').validate({
        rules: {
            bank: {
                required: true
            },
            card: {
                required: true
            },
            discount: {
                required: true
            },
            start_time: {
                required: true
            },
            end_time: {
                required: true
            }
        },
        errorElement: 'span',
        errorPlacement: function(error,el){
            el.parent().find('[data-control="form-input-error"]').append(error);
        },
        submitHandler:function(form) {
            IVYAlert("Loading...", {
                toggle: true
            });

            var data = {
                id: form.id.value,
                bank: form.bank.value,
                card: form.card.value,
                discount: form.discount.value,
                start_time: form.start_time.value,
                end_time: form.end_time.value
            };

            $.post('/cms/index.php/discount/do_modify', data, function(rs){
                if(rs.ret == 0){
                    IVYAlert(rs.msg, function(){
                        location.href = "/cms/index.php/discount/index";
                    });
                }else{
                    IVYAlert(rs.msg);
                }
            },'json');
        }
    });
})
