// build-task/test/proj/resource/page/app/banner/js/banner.modify.swf.config.js
/**
 * @author
 * @desc
 */
define(function(require, exports, module){
    var $ = require('jquery');
    var IVYAlert = require('ivy.alert');

    return {
        file_queue_error_handler: function(file, errorCode, msg) {
            switch (errorCode){
                case SWFUpload.QUEUE_ERROR.QUEUE_LIMIT_EXCEEDED:
                    IVYAlert('Single image only!');
                    break;
                case SWFUpload.QUEUE_ERROR.ZERO_BYTE_FILE:
                    IVYAlert('File size is 0!');
                    break;
                case SWFUpload.QUEUE_ERROR.FILE_EXCEEDS_SIZE_LIMIT:
                    IVYAlert('The image exceeds 2M.');
                    break;
                case SWFUpload.QUEUE_ERROR.INVALID_FILETYPE:
                    IVYAlert('Image files can be .bmp, .png, .jpeg, .jpg or .gif');
                    break;
                default:
                    IVYAlert(msg);
                    break;
            }
        },

        file_dialog_complete_handler: function(numFilesSelected, numFilesQueued) {
            if (numFilesQueued > 0) {
                this.startUpload();
                IVYAlert('Uploading...');
            }
        },

        upload_progress_handler: function(file, bytesComplete, totalBytes){

        },

        upload_error_handler: function(file, errorCode, msg){
            IVYAlert(msg);
        },

        upload_success_handler: function(file, rs){
            try{
                rs = JSON.parse(rs);
                if(rs.ret == 0){
                    $('#banner_url').val(rs.data.pic_url);
                    $('#upload_img').show().html('<img src="' + rs.data.pic_url + '"/>');
                    IVYAlert('Upload successfully!');
                }else{
                    IVYAlert(rs.data.err_info.msg);
                }
            }catch(e){
                IVYAlert(rs);
            }
        },

        upload_complete_handler: function(){
        }

    }
})
