// build-task/test/proj/resource/page/app/banner/js/banner.modify.js
/**
 * @author
 * @desc
 */
define(function(require, exports, module){
	var $ = require('jquery');
	var SWFUpload = require('swfupload');
	var swfConfig = require('page/app/banner/js/banner.modify.swf.config');
	var IVYAlert = require('ivy.alert');
	require('validation');

	new SWFUpload({
		// Backend Settings
		upload_url: "/cms/index.php/upload/up2tfs",
		post_params: {
			/*'width': 560,
			'height': 170,
			'check_size': true*/
		},

		// File Upload Settings
		file_size_limit : "2 MB",	// 2MB
		file_types : "*.jpg;*.png;*.gif",
		file_types_description : "JPG/PNG/GIF Images",
		file_upload_limit : "0",

		file_queue_error_handler : swfConfig.file_queue_error_handler,
		file_dialog_complete_handler : swfConfig.file_dialog_complete_handler,
		upload_progress_handler : swfConfig.upload_progress_handler,
		upload_error_handler : swfConfig.upload_error_handler,
		upload_success_handler : function(file, rs){
			try{
				rs = JSON.parse(rs);
				if(rs.ret == 0){
					$('#banner_url').val(rs.data.pic_url);
					$('#upload_img').removeClass('hidden').attr('src', rs.data.pic_url);
					IVYAlert('Upload successfully!');
				}else{
					IVYAlert(rs.data.err_info.msg);
				}
			}catch(e){
				IVYAlert(rs);
			}
		},
		upload_complete_handler : swfConfig.upload_complete_handler,

		button_placeholder_id : "upload",
		button_width: 1000,
		button_height: 300,
		button_window_mode: SWFUpload.WINDOW_MODE.TRANSPARENT,
		button_cursor: SWFUpload.CURSOR.HAND,

		// Flash Settings
		flash_url : "/cms/resource/lib/cmd/swf/swfupload.swf",

		// Debug Settings
		debug: false
	});

	// tsv
	new SWFUpload({
		// Backend Settings
		upload_url: "/cms/index.php/upload/do_upload",
		post_params: {
			'user_name': 's_id'
		},

		// File Upload Settings
		file_size_limit : "2 MB",	// 2MB
		file_types : "*.tsv",
		file_types_description : "tsv",
		file_upload_limit : "0",

		file_queue_error_handler : function(file, errorCode, msg) {
			switch (errorCode){
				case SWFUpload.QUEUE_ERROR.QUEUE_LIMIT_EXCEEDED:
					IVYAlert('Single file only!');
					break;
				case SWFUpload.QUEUE_ERROR.ZERO_BYTE_FILE:
					IVYAlert('File size is 0!');
					break;
				case SWFUpload.QUEUE_ERROR.FILE_EXCEEDS_SIZE_LIMIT:
					IVYAlert('The file exceeds 2M.');
					break;
				case SWFUpload.QUEUE_ERROR.INVALID_FILETYPE:
					IVYAlert('Files only can be .tsv.');
					break;
				default:
					IVYAlert(msg);
					break;
			}
		},
		file_dialog_complete_handler : swfConfig.file_dialog_complete_handler,
		upload_progress_handler : swfConfig.upload_progress_handler, //
		upload_error_handler : swfConfig.upload_error_handler,
		upload_success_handler : function(file, rs){
			try{
				rs = JSON.parse(rs);
				if(rs.ret == 0){
					$('#tsv_file').val(rs.data.upload_path);
					$('#tsv_file_name').val('File Name: ' + rs.data.upload_path.split('/')[1]);
					IVYAlert('Upload successfully!');
				}else{
					IVYAlert(rs.data.err_info.msg);
				}
			}catch(e){
				IVYAlert(rs);
			}
		},
		upload_complete_handler : swfConfig.upload_complete_handler,

		button_placeholder_id : "tsv",
		button_width: 1000,
		button_height: 300,
		button_window_mode: SWFUpload.WINDOW_MODE.TRANSPARENT,
		button_cursor: SWFUpload.CURSOR.HAND,

		// Flash Settings
		flash_url : "/cms/resource/lib/cmd/swf/swfupload.swf",

		// Debug Settings
		debug: false
	});


	function getRuleInfoData(form){
		var type = form.rule_type.value;
		if(type == 1){
			var val = form.rules.value;
			if(val === ""){
				return "Rules is required!";
			}else{
				return {
					rules: val,
					sorts: {
						sort_by: form.sorts.value,
						sort_type: form.sort_type.value
					}
				}
			}
		}else if(type == 2){
			if(form.tsv_file.value === ""){
				return "Tsv file is required";
			}else{
				return {
					tsv_file: form.tsv_file.value
				}
			}
		}else if(type == 3){
			if(form.action_url.value === ""){
				return "Action url is required";
			}else{
				return {
					action_url: form.action_url.value
				}
			}
		}else if(type == 4){
			if(form.url.value === ""){
				return "Url is required";
			}else{
				return {
					url: form.url.value
				}
			}
		}else{
			return false;
		}
	}
	$('#rule_type').change(function(){
		$('.form-panel').addClass('hidden');
		$('#type' + this.value).removeClass('hidden');
	}).change();


	// form
	$('#form').validate({
		rules:{
			title: {
				required: true
			}
		},
		errorElement: 'span',
		errorPlacement: function(error,el){
			el.parent().find('[data-control="form-input-error"]').append(error);
		},
		submitHandler:function(form) {
			if(!$('#banner_url').val()){
				IVYAlert('Banner is required!');
				return;
			}
			IVYAlert("Loading...", {
				toggle: true
			});
			var data = {
				id: form.id.value,
				title: form.title.value,
				banner_url: form.banner_url.value,
				rule_type: form.rule_type.value
			}

			var rule_info = getRuleInfoData(form);
			if(typeof rule_info === "string"){
				alert(rule_info);
				return;
			}
			data.rule_info = JSON.stringify(rule_info);

			$.post('/cms/index.php/banner/do_modify', data, function(rs){
				if(rs.ret == 0){
					IVYAlert(rs.msg, function(){
						location.href = "/cms/index.php/banner/index";
					});
				}else{
                    IVYAlert(rs.msg);
                }
			},'json');
		}
	});
})
