// build-task/test/proj/resource/page/app/banner/js/banner.list.js
/**
 * @author
 * @desc
 */
define(function(require, exports, module){
	var $ = require('jquery');
	var IVYAlert = require('ivy.alert');
	var Pagingbar = require('pagingbar');
	require('ipick.banner-preview');

	
	$('#modified-preview, #live-preview').bannerPreview();

	// Rewrite goto
	Pagingbar.gotoPage = function(pnum){
		if(pnum > 0 && pnum <= Pagingbar._data.total_page_num){
			Pagingbar.setPageNum(pnum);
			// Set index ui
			$('#common-pagingbar').find('li').removeClass('active');
			$('#common-pagingbar').find('a[data-val="' + pnum + '"]').parent().addClass('active');
			$.get('/cms/index.php/banner/get_more', {
				page: Pagingbar.getPageNum(),
				num : Pagingbar.getPageSize()
			}, function(rs){
				var list = renderList(rs.data.data.list);
				if(list !== ""){
					$('#homeList').children().remove();
					$('#homeList').html(list);
				}
			},'json')
		}else{
			$('#common-pagingbar-goto').focus();
		}
	}
	// Init paging
	Pagingbar.init({
		page_size:  parseInt($('#common-pagingbar-page_size').val()),
		page_num:   parseInt($('#common-pagingbar-page_num').val()),
		total_page_num: parseInt($('#common-pagingbar-total_page_num').val()),
		country: $('#country').val()
	});
	/**
	 * 刷新列表
	 */
	var tplListData = require('page/app/banner/js/banner.list.data.js');
	var _bannerType = $('#bannerTypeList').val().split(',');
	var bannerType = {};
	var tmp;
	for(var i = 0,l = _bannerType.length; i < l;i++){
		tmp = _bannerType[i].split('|');
		bannerType[tmp[0]] = tmp[1];
	}
	function renderList(data){
		if(!data || data.length === 0){
			IVYAlert('No data');
			return "";
		}else{
			var html = "";
			var _idx = (Pagingbar.getPageNum() - 1) * Pagingbar.getPageSize() + 1;
			for(var i = 0, l = data.length;i < l;i++){
				html += tplListData.replace(/{index}/g, _idx + i).
						replace(/{title}/g, data[i].title).
						replace(/{banner_url}/g, data[i].banner_url).
						replace(/{banner_type}/g, bannerType[data[i].rule_type]).
						replace(/{status_cls}/g, data[i].status == 1 ? 'color_green' : 'color_red').
						replace(/{status_val}/g, data[i].status == 1 ? 1 : 0).
						replace(/{status_title}/g, data[i].status == 1 ? 'Click to offline' : 'Click to online').
						replace(/{status_icon}/g, data[i].status == 1 ? 'glyphicon-ok-sign' : 'glyphicon-minus-sign"').
						replace(/{last_update}/g, data[i].last_update).
						replace(/{id}/g, data[i]._id_h).
						replace(/{show_index}/g, data[i].show_index);
			}
			return html;
		}
	}

	// Change Status
	$('#homeList').delegate('a[data-rel="status"]', 'click', function(){
		var $this = $(this);
		if($this.attr('data-val') == "0"){
			$.getJSON('/cms/index.php/banner/online', {
				id: $this.attr('data-id')
			}, function(rs){
				if(rs.ret == 0){
					// Change itself
					$this.attr('data-val', 1).attr('title', 'Click to offline').removeClass('color_red').addClass('color_green').children().removeClass('glyphicon-minus-sign').addClass('glyphicon-ok-sign');

					// Render
					$('#home_online').html(renderPreview(rs.data.online_list));
				}else{
					IVYAlert(rs.msg);
				}
			})
		}else{
			$.getJSON('/cms/index.php/banner/offline', {
				id: $this.attr('data-id')
			}, function(rs){
				if(rs.ret == 0){
					// Change itself
					$this.attr('data-val', 0).attr('title', 'Click to online').removeClass('color_green').addClass('color_red').children().removeClass('glyphicon-ok-sign').addClass('glyphicon-minus-sign');

					// Render
					$('#home_online').html(renderPreview(rs.data.online_list));
				}else{
					IVYAlert(rs.msg);
				}
			})
		}
	});

	// Release
	$('#release-btn').click(function(){
		$.getJSON('/cms/index.php/release/banner', "", function(rs){
			if(rs.ret == 0){
				IVYAlert(rs.msg, function(){
					$('#home_release').html($('#home_online').html());
				})
			}else{
				IVYAlert(rs.msg);
			}
		});
	})


	function renderPreview(data){
		if(!data){
			return "<li></li><li></li><li></li><li></li><li></li><li></li>";
		}else{
			var tpl = '<li>\
						<div>\
							<img src="{$banner_url}">\
							<p>{$title}</p>\
						</div>\
					</li>';
			var rs = [];
			for(var i = 0,l = data.length;i < l;i++){
				rs.push(tpl.replace('{$banner_url}', data[i].banner_url).replace('{$title}', data[i].title));
			}
			for(i = 6 - l;i > 0;i--){
				rs.push('<li></li>');
			}
			return rs.join("");
		}
	}

	// Change show_index
	var SPage = {
		sortEvent: function(){
			$('#homeList').delegate('input[data-rel="sort"]', 'change', function(){
				var val = +parseInt(this.value);
				if(isNaN(val)){
					return;
				}
				var $this = $(this);
				$.getJSON('/cms/index.php/banner/modify_index', {
					id: $this.attr('data-id'),
					show_index: val
				}, function(rs){
					$('#home_online').html(renderPreview(rs.data.online_list));
				})
			})
		}
	};
	SPage.sortEvent();
})
