// build-task/test/proj/resource/page/app/banner/js/modify.js

define(function (require, exports, module) {

	var $ = require('jquery');
	require('jquery-validate');
	require('jquery.form');
	var IVYAlert = require('ivy.alert');

	require('ipick.file-upload');

	// Upload Image
	$('#image-upload').fileUpload({
		url: '/cms/index.php/upload/up2tfs',
		/*params: {
			'width': 560,
			'height': 170,
			'check_size': true
		},*/
		file: {
			types: '*.jpg;*.png;*.gif',
			typesDescription: 'JPG/PNG/GIF Images',
			sizeLimit: '2 MB',
		},
	}).on('uploadsuccess', function (event) {
		var data = $.parseJSON(event.body);
		if (data['ret']) {
			IVYAlert(data['msg']);
		} else {
			$('#image').val(data['data']['pic_url']);
			$('#uploaded-image').removeClass('hidden').attr('src', data['data']['pic_url']);
			IVYAlert('Image uploaded successfully');
		}
	}).on('uploadstart', function (event) {
		IVYAlert('Uploading...', { toggle: true });
	}).on('error', function (event) {
		var message = {
			'QUEUE_ERROR.QUEUE_LIMIT_EXCEEDED': 'You can only upload one file at a time',
			'QUEUE_ERROR.ZERO_BYTE_FILE': 'The file is empty',
			'QUEUE_ERROR.FILE_EXCEEDS_SIZE_LIMIT': 'The file exceeded the maximum file size (2 MB)',
			'QUEUE_ERROR.INVALID_FILETYPE': 'The file format should be .bmp, .png, .jpeg, .jpg or .gif'
		}[event.error.type] || event.error.message;
		IVYAlert(message);
	});

	// Upload TSV File
	$('#tsv-upload').fileUpload({
		url: '/cms/index.php/upload/do_upload',
		file: {
			types: '*.tsv',
			typesDescription: 'TSV Files',
			sizeLimit: '2 MB',
		},
	}).on('uploadsuccess', function (event) {
		var data = $.parseJSON(event.body);
		if (data['ret']) {
			IVYAlert(data['msg']);
		} else {
			$('#tsv').val(data['data']['upload_path']);
			$('#tsv-name').val('File Name: ' + data['data']['upload_path'].replace(/^.*\//, ''));
			IVYAlert('TSV file uploaded successfully');
		}
	}).on('uploadstart', function (event) {
		IVYAlert('Uploading...', { toggle: true });
	}).on('error', function (event) {
		var message = {
			'QUEUE_ERROR.QUEUE_LIMIT_EXCEEDED': 'You can only upload one file at a time',
			'QUEUE_ERROR.ZERO_BYTE_FILE': 'The file is empty',
			'QUEUE_ERROR.FILE_EXCEEDS_SIZE_LIMIT': 'The file exceeded the maximum file size (2 MB)',
			'QUEUE_ERROR.INVALID_FILETYPE': 'The file format should be .tsv'
		}[event.error.type] || event.error.message;
		IVYAlert(message);
	});

	// Switch Type
	$('#type').change(function (event) {
		var activePanel = {
			'1': $('#type-tags-panel'),
			'4': $('#type-url-panel'),
			'3': $('#type-app-panel'),
			'2': $('#type-tsv-panel')
		}[$(this).val()];
		$('.form-panel').addClass('hidden');
		if (activePanel) activePanel.removeClass('hidden');
	}).change();

	// Validation
	$('#form').validate({
		errorPlacement: function (error, element) {
		}
	});

	// Ajax Submit
	$('form')
		.ajaxForm({
			beforeSerialize: function (data) {
				if(!$('#image').val()){
					IVYAlert('You should upload an image before submit');
					return false;
				}
				if ($('#type').val() === '1') {
					$('#rule').val(JSON.stringify({
						'rules': $('#rules').val(),
						'sorts': {
							'sort_by': $('#sort-key').val(),
							'sort_type': $('#sort-order').val()
						}
					}));
				}
				if ($('#type').val() === '4') {
					$('#rule').val(JSON.stringify({
						'url': $('#url').val()
					}));
				}
				if ($('#type').val() === '3') {
					$('#rule').val(JSON.stringify({
						'action_url': $('#action-url').val()
					}));
				}
				if ($('#type').val() === '2') {
					if(!$('#tsv').val()){
						IVYAlert('You should upload an TSV file before submit');
						return false;
					}
					$('#rule').val(JSON.stringify({
						'tsv_file': $('#tsv').val()
					}));
				}
			},
			beforeSubmit: function (data) {
				IVYAlert('Submitting...', { toggle: true });
			},
			success: function (data) {
				data = $.parseJSON(data);
				if (data['ret']) {
					IVYAlert(data['msg']);
				} else {
					IVYAlert('Your edit has been submitted', function () {
						window.location = '../banner/index';
					});
				}
			}
		});

});
