// build-task/test/proj/resource/page/app/price/price.modify.js
/**
 * @author
 * @desc
 */
define(function(require, exports, module){
    var $ = require('jquery');
    var IVYAlert = require('ivy.alert');
    require('validation');
    var IPICKFclangname = require('ipick.fc.langname');

    // 初始化需要多语言添加的表单项
    var category = new IPICKFclangname();
    category.init('#category', $('#category').children().children(),{
        add_btn_txt: 'Add Category',
        data: $('#category').attr('data-val') ? JSON.parse($('#category').attr('data-val')) : []
    });

    var minPrice = new IPICKFclangname();
    minPrice.init('#min_price', $('#min_price').children().children(),{
        add_btn_txt: 'Add Price',
        data: $('#min_price').attr('data-val') ? JSON.parse($('#min_price').attr('data-val')) : []
    });

    var maxPrice = new IPICKFclangname();
    maxPrice.init('#max_price', $('#max_price').children().children(),{
        add_btn_txt: 'Add Price',
        data: $('#max_price').attr('data-val') ? JSON.parse($('#max_price').attr('data-val')) : []
    });

    $('#form').validate({
        rules: {
            order: {
                required: true
            },
            main_order: {
                required: true
            }
        },
        errorElement: 'span',
        errorPlacement: function(error,el){
            el.parent().find('[data-control="form-input-error"]').append(error);
        },
        submitHandler:function(form) {
            IVYAlert("Loading...", {
                toggle: true
            });

            var data = {
                id: form.id.value,
                order: form.order.value,
                category_value: category.getData(),
                min_price: minPrice.getData(),
                max_price: maxPrice.getData()
            };

            if(data.max_price === ""){
                IVYAlert('Max Price is required!');
                return;
            }

            $.post('/cms/index.php/price/do_modify', data, function(rs){
                if(rs.ret == 0){
                    IVYAlert(rs.msg, function(){
                        location.href = "/cms/index.php/price/index";
                    });
                }else{
                    IVYAlert(rs.msg);
                }
            },'json');
        }
    });
})
