// build-task/test/proj/resource/page/app/landmark/js/landmark.list.js
/**
 * @author
 * @desc
 */
define(function(require, exports, module){
    var $ = require('jquery');
    var IVYAlert = require('ivy.alert');
    var Pagingbar = require('pagingbar');

    // Rewrite goto
    Pagingbar.gotoPage = function(pnum){
        if(pnum > 0 && pnum <= Pagingbar._data.total_page_num){
            Pagingbar.setPageNum(pnum);
            // Set index ui
            $('#common-pagingbar').find('li').removeClass('active');
            $('#common-pagingbar').find('a[data-val="' + pnum + '"]').parent().addClass('active');
            $.get('/cms/index.php/landmark/get_more', {
                page: Pagingbar.getPageNum(),
                num : Pagingbar.getPageSize()
            }, function(rs){
                var list = renderList(rs.data.data.list);
                if(list !== ""){
                    $('#homeList').children().remove();
                    $('#homeList').html(list);
                }
            },'json')
        }else{
            $('#common-pagingbar-goto').focus();
        }
    }
    // Init paging
    Pagingbar.init({
        page_size:  parseInt($('#common-pagingbar-page_size').val()),
        page_num:   parseInt($('#common-pagingbar-page_num').val()),
        total_page_num: parseInt($('#common-pagingbar-total_page_num').val()),
        country: $('#country').val()
    });
    /**
     * 刷新列表
     */
    var tplListData = require('page/app/landmark/js/landmark.list.data.js');
    function renderList(data){
        if(!data || data.length === 0){
            IVYAlert('No data');
            return "";
        }else{
            var html = "";
            var _idx = (Pagingbar.getPageNum() - 1) * Pagingbar.getPageSize() + 1;
            for(var i = 0, l = data.length;i < l;i++){
                html += tplListData.replace(/{index}/g, _idx + i).
                replace(/{name}/g, data[i].name).
                replace(/{ycoordinate}/g, data[i].ycoordinate).
                replace(/{xcoordinate}/g, data[i].xcoordinate).
                replace(/{radius}/g, data[i].radius).
                replace(/{alias}/g, data[i].alias).
                replace(/{id}/g, data[i].id);
            }
            return html;
        }
    }


    // 删除功能
    $('#homeList').delegate('a[data-rel="delete"]', 'click', function(){
        var $this = $(this);
        if(confirm('Are you sure?')){
            $.getJSON('/cms/index.php/landmark/delete', {
                id: $this.attr('data-val')
            }, function(rs){
                if(rs.ret === 0){
                    $('#common-pagingbar-goto').next().click();
                }else{
                    IVYAlert(rs.msg);
                }
            })
        }
    });


    // 搜索功能
    $('#search').click(function(){
        var txt = $.trim($(this).parent().prev().val());
        if(txt === ''){
            location.reload();
            return;
        }
        $.get('/cms/index.php/landmark/search', {
            page: 1,
            num : Pagingbar.getPageSize(),
            keyword: txt
        }, function(rs){
            var list = renderList(rs.data.data.all_list);
            if(list !== ""){
                $('#homeList').children().remove();
                $('#homeList').html(list);
            }
            $('#common-pagingbar').hide();
        },'json')
    })

})
