// build-task/test/proj/resource/page/app/landmark/js/landmark.modify.js
/**
 * @author
 * @desc
 */
define(function(require, exports, module){
    var $ = require('jquery');
    var IVYAlert = require('ivy.alert');
    require('validation');

    $('#form').validate({
        rules: {
            landmark_en: {
                required: true
            },
            landmark_cn: {
                required: true
            },
            ycoordinate: {
                required: true
            },
            xcoordinate: {
                required: true
            },
            radius: {
                required: true
            },
            alias: {
                required: true
            },
            region_rank: {
                required: true
            }
        },
        errorElement: 'span',
        errorPlacement: function(error,el){
            el.parent().find('[data-control="form-input-error"]').append(error);
        },
        submitHandler:function(form) {
            IVYAlert("Loading...", {
                toggle: true
            });
            var data = {
                id: form.id.value,
                landmark_en: form.landmark_en.value,
                landmark_cn: form.landmark_cn.value,
                ycoordinate: form.longitude.value,
                xcoordinate: form.latitude.value,
                radius: form.radius.value,
                alias: form.alias.value,
                region_rank: form.region_rank.value
            };

            $.post('/cms/index.php/landmark/domodify', data, function(rs){
                if(rs.ret == 0){
                    IVYAlert(rs.msg, function(){
                        location.href = "/cms/index.php/landmark/index";
                    });
                }else{
                    IVYAlert(rs.msg);
                }
            },'json');
        }
    });
})
