// build-task/test/proj/resource/page/app/detail/js/detail.modify.js
/**
 * @author
 * @desc
 */
define(function(require, exports, module){
    var $ = require('jquery');
    var SWFUpload = require('swfupload');
    var IVYAlert = require('ivy.alert');
    require('validation');

    // New upload control
    new SWFUpload({
        // Backend Settings
        upload_url: "/cms/index.php/upload/up2tfs",

        // File Upload Settings
        file_size_limit : "2 MB",	// 2MB
        file_types : "*.jpg;*.png;*.gif",
        file_types_description : "JPG/PNG/GIF Images",
        file_upload_limit : "0",

        file_queue_error_handler : function(file, errorCode, msg) {
            switch (errorCode){
                case SWFUpload.QUEUE_ERROR.QUEUE_LIMIT_EXCEEDED:
                    alert('Single image only!');
                    break;
                case SWFUpload.QUEUE_ERROR.ZERO_BYTE_FILE:
                    alert('File size is 0!');
                    break;
                case SWFUpload.QUEUE_ERROR.FILE_EXCEEDS_SIZE_LIMIT:
                    alert('The image exceeds 2M.');
                    break;
                case SWFUpload.QUEUE_ERROR.INVALID_FILETYPE:
                    alert('Image files can be .bmp, .png, .jpeg, .jpg or .gif');
                    break;
                default:
                    alert(msg);
                    break;
            }
        },
        file_dialog_complete_handler : function(numFilesSelected, numFilesQueued) {
            if (numFilesQueued > 0) {
                this.startUpload();
            }
        },
        upload_progress_handler : function(file, bytesComplete, totalBytes){
        },
        upload_error_handler : function(file, errorCode, msg){
            IVYAlert(msg);
        },
        upload_success_handler : function(file, rs){
            try{
                rs = JSON.parse(rs);
                if(rs.ret == 0){
                    var $list = $('#imgList').children();
                    var which = '';
                    if($list.length === 0){
                        which = 'first last';
                    }else{
                        which = 'last';
                    }
                    $list.last().removeClass('last');
                    $('<li class="{which}">\
                            <img src="{pic_url}" data-val="{pic_url}">\
                            <div class="covers_fns">\
                                <span class="cover" data-rel="cover" title="設置為封面">⊙</span>\
                                <span class="prev" data-rel="prev" title="前移"><<</span>\
                                <span class="next" data-rel="next" title="後移">>></span>\
                                <span class="del" data-rel="del" title="刪除">×</span>\
                            </div>\
                            <div class="covers_idx" data-rel="index">{index}</div>\
                        </li>'.replace('{which}', which).replace(/\{pic_url\}/g, rs.data.pic_url).replace('{index}', $list.length + 1)).appendTo($('#imgList'));
                    IVYAlert('Upload successfully!');
                }else{
                    IVYAlert(rs.msg);
                }
            }catch(e){
                IVYAlert(rs);
            }
        },
        upload_complete_handler : function(){},

        button_placeholder_id : "upload",
        button_width: 161,
        button_height: 30,
        button_text_top_padding: 5,
        button_window_mode: SWFUpload.WINDOW_MODE.TRANSPARENT,
        button_cursor: SWFUpload.CURSOR.HAND,

        wrapStyle: 'position: relative;height: 30px;width: 160px;',
        objectStyle: 'position: absolute;left: 0;z-index:9',
        btnElement: '<button type="button" class="btn-upload" style="position: absolute;z-index: 8">上傳圖片</button>',

        // Flash Settings
        flash_url : "/cms/resource/lib/cmd/swf/swfupload.swf",

        // Debug Settings
        debug: false
    });

    // All of events refer to the form
    var FormEvent = {
        init: function(){
            this.sort();
            this.delImg();
            this.checkItem();
            this.rating();
            this.get2Region();
            this.get2Dishes();
            this.getLandmarkList();
            this.addDiscount();
        },

        /**
         * 餐厅图片排序
         */
        sort: function(){
            $('#imgList').delegate('span[data-rel="cover"]', 'click', function(){
                var $this = $(this).parent().parent();
                var $replace = $this.parent().children().eq(0).find('img');
                var $current = $this.find('img');
                $current.replaceWith($replace.clone());
                $replace.replaceWith($current);
            });
            $('#imgList').delegate('span[data-rel="prev"]', 'click', function(){
                var $this = $(this).parent().parent();
                var $prev = $this.prev();
                if($prev.length === 0){
                    return;
                }
                var $current = $this.find('img');
                var $replace = $prev.find('img');
                $current.replaceWith($replace.clone());
                $replace.replaceWith($current);
            });
            $('#imgList').delegate('span[data-rel="next"]', 'click', function(){
                var $this = $(this).parent().parent();
                var $next = $this.next();
                if($next.length === 0){
                    return;
                }
                var $current = $this.find('img');
                var $replace = $next.find('img');
                $current.replaceWith($replace.clone());
                $replace.replaceWith($current);
            });
        },

        /**
         * 删除图片
         */
        delImg: function(){
            $('#imgList').delegate('span[data-rel="del"]', 'click', function(){
                var $target = $(this).parent().parent();
                var $wrap = $target.parent();
                var $list = $wrap.children();
                // 判断是不是第一个或最后一个
                if($list.index($target) === 0){
                    $target.next().addClass('first');
                }else if($list.index($target) === $list.length - 1){
                    $target.prev().addClass('last');
                }
                $target.remove();
                $list.each(function(idx){
                    $(this).find('div[data-rel="index"]').text(idx + 1);
                })
            });
        },

        /**
         * 获取图片列表
         */
        getImgList: function(){
            var output = [];
            $('#imgList').find('img').each(function(){
                output.push($(this).attr('data-val'));
            });
            if(output.length === 0){
                return false;
            }else{
                return output;
            }
        },

        /**
         * 检阅
         */
        checkItem: function(){
            $('#checkType').change(function(){
                $.getJSON('/web/index.php/ipickweb/check_item', {
                    restaurantid: $('#id').val(),
                    type: this.checked ? 1 : 0
                }, function(rs){
                    if(rs.ret == 0){
                    }else{
                        alert(rs.msg);
                    }
                });
            })
        },

        /**
         * 评分
         */
        rating: function(){
            function ratingShow(val){
                var offset = $(this).offset();
                $('#ratingShow').text(val).css({
                    top: offset.top - 25,
                    left: offset.left + 5
                }).fadeIn(300,function(){
                    $(this).fadeOut(700)
                });
            }
            $('ul[data-rel="rating"]').children().hover(function(e){
                var $parent = $(this);
                var v = +$parent.attr('data-val');
                $parent = $parent.parent();
                $parent = $parent.children().removeClass('active');
                for(v--;v > -1;v--){
                    $parent.eq(v).addClass('active');
                }
            }, function(){
                var $parent = $(this).parent();
                var v = +$parent.attr('data-val');
                $parent = $parent.children().removeClass('active');
                for(v--;v > -1;v--){
                    $parent.eq(v).addClass('active');
                }
            }).on('click', function(){
                var $this = $(this);
                var $parent = $this.parent();
                var $children = $parent.children();
                var rating = $this.attr('data-val');

                $.getJSON('/cms/index.php/ipickweb/rating', {
                    rating_score: rating,
                    restaurant_id: $parent.attr('data-id')
                }, function(rs){
                    if(rs.ret == 0){
                        $parent.attr('data-val', rating);
                        ratingShow.call($this[0], '+' + rating)
                    }else{
                        $children.removeClass('active');
                    }
                })
            });
        },

        /**
         * 二级地区请求
         */
        get2Region: function(){
            $('#regionF').change(function(){
                $('#regionS').attr('disabled', '');
                $.getJSON('/cms/index.php/region/get_region', {
                    id: this.value
                }, function(rs){
                    if(rs.ret == 0){
                        rs.data.data.region_list ? render2RegionOptions(rs.data.data.region_list) : null;
                    }else{
                        IVYAlert(rs.msg);
                    }
                })
            }).change();

            function render2RegionOptions(data){
                var options = "";
                for(var i = 0,l = data.length;i < l;i++){
                    options += '<option value="' + data[i].id + '" data-name="' + data[i].name + '">' + data[i].name + '</option>'
                }
                var $target = $('#regionS');
                $target.html(options);
                $target.removeAttr('disabled');
                if($target.attr('data-val')){
                    $target.children('[data-name="' + $target.attr('data-val') + '"]').attr('selected', '');
                    $target.removeAttr('data-val');
                }
            }
        },

        /**
         * 二级菜式请求
         */
        get2Dishes: function(){
            $('#subTypeF').change(function(){
                $('#subTypeS').attr('disabled', '');
                $.getJSON('/cms/index.php/cuisine/get_cuisine', {
                    id: this.value
                }, function(rs){
                    if(rs.ret == 0){
                        rs.data.data.cuisine_list ? render2DishesOptions(rs.data.data.cuisine_list) : null;
                    }else{
                        IVYAlert(rs.msg);
                    }
                })
            }).change();

            function render2DishesOptions(data){
                var options = "";
                for(var i = 0,l = data.length;i < l;i++){
                    options += '<option value="' + data[i].id + '" data-name="' + data[i].name + '">' + data[i].name + '</option>'
                }
                var $target = $('#subTypeS');
                $target.html(options);
                $target.removeAttr('disabled');
                if($target.attr('data-val')){
                    $target.children('[data-name="' + $target.attr('data-val') + '"]').attr('selected', '');
                    $target.removeAttr('data-val');
                }
            }
        },

        /**
         * 地标选择
         */
        getLandmarkList: function(){
            // 新增新地标
            $('#addLandmark').click(function(){
                $('#landmarkItem').children().clone().appendTo($('#formLandmarkList'));
            })

            // 编辑地标
            var currentLandmark = null;
            $('#formLandmarkList').delegate('input[data-rel="landmark"]', 'click focus', function(){
                $('#landmarkShow').show();
                currentLandmark = this;
            });
            // 删除地标
            $('#formLandmarkList').delegate('a[data-rel="landmark_item_del"]', 'click', function(){
                $(this).parent().remove();
            });

            // 搜索
            $('#landmarkSearch').next().click(function(){
                var searchtxt = $.trim($(this).prev().val());
                if(searchtxt){
                    $.getJSON('/cms/index.php/landmark/search', {
                        keyword: searchtxt
                    }, function(rs){
                        if(rs.ret == 0){
                            if(rs.data.data.all_list){
                                rs.data.data.all_list.length > 0 ? renderLandmarkList(rs.data.data.all_list) : IVYAlert("無數據！");
                            }else{
                                IVYAlert("無數據！");
                            }
                        }else{
                            IVYAlert(rs.msg);
                        }
                    })
                }
            })

            // 选中某个地标
            $('#landmarkList').delegate('li', 'click', function(){
                var $this = $(this);
                $(currentLandmark).val($this.text()).next().val($this.attr('data-val'));
                $('#landmarkShow').hide();
            });

            // 关闭
            $('#landmarkClose').click(function(){
                $('#landmarkShow').hide();
            })

            function renderLandmarkList(data){
                var html = "";
                for(var i = 0, l = data.length;i < l;i++){
                    html += '<li data-val="' + data[i].id + '">' + data[i].name + '</li>';
                }
                $('#landmarkList').html(html);
            }
        },

        /**
         * 获取地标数据
         */
        getLandmarkData: function(){
            var rs = [];
            $('input[data-name="landmark"]').each(function(){
                if(this.value){
                    rs.push(this.value);
                }
            });
            return rs;
        },


        /**
         * 添加折扣优惠
         */
        addDiscount: function(){
            $('#discountFL').change(function(){
                if(this.value === ""){
                    return;
                }

                $('#discountSL, #discountTL').children().remove();
                $.getJSON('/cms/index.php/discount/get_bank_type_list', {
                    bank_name: $(this).val()
                }, function(rs){
                    if(rs.ret === 0){
                        $('#discountSL, #discountTL').children().remove();
                        $('#discountSL').html(renderSLSelectData(rs.data.card_list));
                    }else{
                        IVYAlert(rs.msg);
                    }
                });
            });

            $('#discountSL').change(function(){
                if(this.value === ""){
                    return;
                }

                $.getJSON('/cms/index.php/discount/get_discount_list', {
                    bank_name: $('#discountFL').val(),
                    card_type: $(this).val()
                }, function(rs){
                    if(rs.ret === 0){
                        $('#discountTL').children().remove();
                        $('#discountTL').html(renderTLSelectData(rs.data.discount_list));
                    }else{
                        IVYAlert(rs.msg);
                    }
                });
            });

            function renderSLSelectData(data){
                if(!data){
                    return '';
                }
                var tpl = '<option value="{v}">{v}</option>'
                var output = '';
                for(var i = 0,l = data.length;i < l;i++){
                    output += tpl.replace(/{v}/g, data[i]);
                }
                return '<option value="">---</option>' + output;
            }

            function renderTLSelectData(data){
                if(!data){
                    return '';
                }
                var tpl = '<option value="{val}">{txt}</option>'
                var output = '';
                for(var itm in data){
                    output += tpl.replace(/{val}/, itm).replace(/{txt}/, data[itm]);
                }
                return output;
            }


            // 添加折扣
            $('#addDiscount').click(function(){
                var tpl = '<li>\
                                <p>{bank}<strong>|</strong>{card}<strong>|</strong>{discount}</p>\
                                <input type="hidden" value="{val}"/>\
                                <a href="javascript:void(0)" data-rel="discount-remove"><span class="glyphicon glyphicon-remove-circle"></span></a>\
                            </li>';
                $(tpl.replace('{bank}', $('#discountFL').val())
                    .replace('{card}', $('#discountSL').val())
                    .replace('{discount}', $('#discountTL').children('[value="' + $('#discountTL').val() + '"]').text())
                    .replace('{val}', $('#discountTL').val()))
                    .appendTo($('#discountList'));
            });
            // 监听删除折扣
            $('#discountList').delegate('a[data-rel="discount-remove"]', 'click', function(){
                $(this).parent().remove();
            })
        },
        getDiscountData: function(){
            var output = [];
            $('#discountList').children().each(function(){
                output.push($(this).find('input').val());
            });
            return output.length === 0 ? "" : JSON.stringify(output);
        }
    }
    FormEvent.init();

    // Form validate
    $('#restaurant-register-form').validate({
        rules:{
            tel: {
                required: true
            }
        },
        errorElement: 'span',
        errorPlacement: function(error,el){
            el.parent().find('[data-control="form-input-error"]').append(error);
        },
        submitHandler:function(form) {
            var picture_list = FormEvent.getImgList();
            if(!picture_list){
                alert('至少需要一張图片！');
                return;
            }

            $.post('/cms/index.php/ipickweb/do_modify', {
                cms_id: form.cms_id.value,
                cuisine_id: $('#subTypeS').val(),
                region_id: $('#regionS').val(),
                landmark_id: JSON.stringify(FormEvent.getLandmarkData()),
                spendingvalue: form.spendingvalue.value,
                restaurantid: form.id.value,
                picture_list: JSON.stringify(picture_list),
                phone: form.phone.value,
                location: form.location.value,
                direction: form.direction.value,
                openinghours: JSON.stringify({
                    openinghour1: form.openinghour1.value,
                    openinghour2: form.openinghour2.value,
                    openinghour3: form.openinghour3.value,
                    openinghour4: form.openinghour4.value,
                    openinghour5: form.openinghour5.value,
                    openinghour6: form.openinghour6.value,
                    openinghour7: form.openinghour7.value
                }),
                discount: FormEvent.getDiscountData()
            }, function(rs){
                if(rs.ret == 0){
                    IVYAlert(rs.msg, function(){
                        history.go(-1);
                    });
                }
            },'json');
        }
    });


    // 回顶部
    $('#navBar').css({
        right: ($(window).width() - 980) / 2 - 100 - 20
    })
    $(window).scroll(function(){
        if($(this).scrollTop() > 0){
            $('#scrollToTop').fadeIn(250);
        }else{
            $('#scrollToTop').fadeOut(500);
        }
    });
    $('#scrollToTop').click(function(){
        $(window).scrollTop(0);
    })
})
