// build-task/test/proj/resource/page/app/detail/js/restsearch.list.js
/**
 * @author
 * @desc
 */
define(function(require, exports, module){
    var $ = require('jquery');
    require('jqueryui');

    $("#keyword").autocomplete({
        minLength: 0,
        source: function( request, response ){
        $.ajax({
            url: "/cms/index.php/ipickweb/keyword_suggest",
            dataType : 'json',
            data: {
            'keyword': request.term
            },
            success: function(ret){
            response( ret.data );
            }
            });
        }
        }).autocomplete( "instance" )._renderItem = function( ul, item ) {
        $("#suggest").append(ul);
        var iconTpl = (item.iconurl == false) ? '' : '<span class="result_icon"><img src="'+item.iconurl+'" alt="" style="width:30px;height:30px;line-height:30px;" /></span>';
        var tpl = '<a href="'+item.search+'">'+iconTpl+'</span><span class="result_title">'+ item.title +'</span><span class="result_count">'+  item.num +'&nbsp;&nbsp;&gt;&gt;</span></a>';

        return $( "<li>" )
        .append(tpl)
        .appendTo( ul );
    };


    // 检阅
    $('#list').delegate('a[data-rel="check"]', 'click', function(){
        var $this = $(this);
        $.getJSON('/cms/index.php/ipickweb/check_item', {
            cms_id: $this.attr('data-cid'),
            restaurantid: $this.attr('data-val'),
            type: $this.attr('data-type')
        }, function(rs){
            if(rs.ret == 0){
                if($this.attr('data-type') == 0){
                    $this.parent().removeClass('list_check_ok').addClass('list_check_no');
                    $this.attr('data-type', 1);
                    $this.attr('title', '未評');
                    $this.children().replaceWith('<icon class="icon_checked_no">≡</icon>');
                }else{
                    $this.parent().removeClass('list_check_no').addClass('list_check_ok');
                    $this.attr('data-type', 0);
                    $this.attr('title', '已評過');
                    $this.children().replaceWith('<icon class="icon_checked_ok">✔</icon>');
                }
            }else{
                IVYAlert(rs.msg);
            }
        });
    })


    // 分页
    $('#prev').click(function(){
        if(+$('#pagingIdx').val() > 1){
            var _href = location.href;
            _href = _href.replace(/&page=[0-9]+/g, '').replace(/\?page=[0-9]+/g, '?');
            var _c;
            if(_href.indexOf('?') > -1){
                _c = '&';
            }else{
                _c = '?';
            }
            location.href = _href + _c + 'page=' + (+$('#pagingIdx').val() - 1);
        }
    });
    $('#next').click(function(){
        if(+$('#pagingIdx').val() < +$('#pagingTotal').val()){
            var _href = location.href;
            _href = _href.replace(/&page=[0-9]+/g, '').replace(/\?page=[0-9]+/g, '?');
            var _c;
            if(_href.indexOf('?') > -1){
                _c = '&';
            }else{
                _c = '?';
            }
            location.href = _href + _c + 'page=' + (+$('#pagingIdx').val() + 1);
        }
    });


    // 回顶部
    $('#navBar').css({
        right: ($(window).width() - 980) / 2 - 100 - 20
    })
    $(window).scroll(function(){
        if($(this).scrollTop() > 0){
            $('#scrollToTop').fadeIn(250);
        }else{
            $('#scrollToTop').fadeOut(500);
        }
    });
    $('#scrollToTop').click(function(){
        $(window).scrollTop(0);
    })
})
