// build-task/test/proj/resource/page/app/common/js/common.pagingbar.js
/**
 * @author pakinguo
 */
define(function(require, exports, module){
    var Pagingbar = {
        _data: {
            page_size: 20,
            page_num: 1,
            total_page_num: 1
        },


        init: function(params){
            if(typeof params !== 'object'){
                return;
            }
            $.extend(Pagingbar._data, params);
            Pagingbar.bindIndexEvt();
            Pagingbar.bindGotoEvt();
            Pagingbar.bindPrevAndNextEvt();
        },


        setPageSize: function(n){
            n = parseInt(n);
            if(isNaN(n)){
                return;
            }
            Pagingbar._data.page_size = n;
        },
        getPageSize: function(){
            return Pagingbar._data.page_size;
        },


        setPageNum: function(n){
            n = parseInt(n);
            if(isNaN(n)){
                return;
            }
            Pagingbar._data.page_num = n;
        },
        getPageNum: function(){
            return Pagingbar._data.page_num;
        },


        bindIndexEvt: function(){
            $('a[data-rel="common-paging-idx"]').click(function(){
                var $this = $(this);
                var pnum = parseInt($this.attr('data-val'));
                Pagingbar.gotoPage(pnum);
                $this.parent().parent().children().removeClass('active');
                $this.parent().addClass('active');
            })
        },


        bindGotoEvt: function(){
            var i;
            $('#common-pagingbar-goto').focus(function(){
                clearTimeout(i);
                $(this).next().fadeIn();
            }).blur(function(){
                var _this = this;
                i = setTimeout(function(){
                    $(_this).next().fadeOut();
                },500);
            }).next().click(function(){
                var pnum = parseInt($(this).prev().val());
                Pagingbar.gotoPage(pnum);
            })
        },


        bindPrevAndNextEvt: function(){
            $('#common-pagingbar-prev').click(Pagingbar.prevPage);
            $('#common-pagingbar-next').click(Pagingbar.nextPage);
        },


        gotoPage: function(pnum){
            if(pnum > 0 && pnum <= Pagingbar._data.total_page_num){
                if(location.search){
                    location.href = location.href.replace(location.search, '?' + $.param($.extend({}, Pagingbar._data, {
                        page_num: pnum
                    })))
                }else{
                    location.href = location.href + '?' + $.param($.extend({}, Pagingbar._data, {
                        page_num: pnum
                    }))
                }
            }else{
                $('#common-pagingbar-goto').focus();
            }
        },


        prevPage: function(){
            if(Pagingbar._data.page_num === 1){
                alert('This is the first page!');
            }else{
                Pagingbar.gotoPage(Pagingbar._data.page_num - 1);
            }
        },


        nextPage: function(){
            if(Pagingbar._data.page_num === Pagingbar._data.total_page_num){
                alert('This is the last page!');
            }else{
                Pagingbar.gotoPage(Pagingbar._data.page_num + 1);
            }
        }
    }

    module.exports = Pagingbar;
})
