// build-task/test/proj/resource/page/app/region/js/region.list.js
/**
 * @author
 * @desc
 */
define(function(require, exports, module){
    var $ = require('jquery');
    var IVYAlert = require('ivy.alert');
    var Pagingbar = require('pagingbar');

    // Rewrite goto
    Pagingbar.gotoPage = function(pnum){
        if(pnum > 0 && pnum <= Pagingbar._data.total_page_num){
            Pagingbar.setPageNum(pnum);
            // Set index ui
            $('#common-pagingbar').find('li').removeClass('active');
            $('#common-pagingbar').find('a[data-val="' + pnum + '"]').parent().addClass('active');
            $.get('/cms/index.php/region/get_more', {
                page: Pagingbar.getPageNum(),
                num : Pagingbar.getPageSize()
            }, function(rs){
                var list = renderList(rs.data.data.list);
                if(list !== ""){
                    $('#homeList').children().remove();
                    $('#homeList').html(list);
                }
            },'json')
        }else{
            $('#common-pagingbar-goto').focus();
        }
    }
    // Init paging
    Pagingbar.init({
        page_size:  parseInt($('#common-pagingbar-page_size').val()),
        page_num:   parseInt($('#common-pagingbar-page_num').val()),
        total_page_num: parseInt($('#common-pagingbar-total_page_num').val()),
        country: $('#country').val()
    });
    /**
     * 刷新列表
     */
    var tplListData = require('page/app/region/js/region.list.data.js');
    function renderList(data){
        if(!data || data.length === 0){
            IVYAlert('No data');
            return "";
        }else{
            var html = "";
            var _idx = (Pagingbar.getPageNum() - 1) * Pagingbar.getPageSize() + 1;
            for(var i = 0, l = data.length;i < l;i++){
                html += tplListData.replace(/{index}/g, _idx + i)
                    .replace(/{name}/g, data[i].name)
                    .replace(/{last_update}/g, data[i].last_update)
                    .replace(/{alias}/g, data[i].alias)
                    .replace(/{id}/g, data[i].id)
            }
            return html;
        }
    }

    // 查看二级列表
    $('#previewSecondLevelCuisineClose').click(function(){
        $('#previewSecondLevelCuisine').hide();
    })
    $('#homeList').delegate('a[data-rel="slevel-preview"]', 'click', function(){
        var $this = $(this);
        $.getJSON('/cms/index.php/region/get_region', {
            id: $this.attr('data-id')
        }, function(rs){
            if(rs.ret == 0){
                // 填充列表
                $('#sLevelCuisineList').html(renderSlevelList(rs.data.data.region_list));
                $('#previewSecondLevelCuisine').show();
                $('#parentName').text($this.attr('data-name'));
            }else{
                IVYAlert(rs.msg);
            }
        })
    });
    var tplDataList = require('page/app/region/js/region.list.s.data.js');
    function renderSlevelList(data){
        if(!data){
            return "";
        }
        var html = "";
        var _idx = 1;
        for(var i = 0, l = data.length;i < l;i++){
            html += tplDataList.replace(/{index}/g, _idx + i)
                    .replace(/{name}/g, data[i].name)
                    .replace(/{last_update}/g, data[i].last_update)
                    .replace(/{alias}/g, data[i].alias)
                    .replace(/{id}/g, data[i].id)
        }
        return html;
    }

})
