// build-task/test/proj/resource/page/app/msg_push/msg.push.js
/**
 * @author
 * @desc
 */
define(function(require, exports, module){
    var $ = require('jquery');
    require('validation');
    var IVYAlert = require('ivy.alert');

    $('#addParam').click(function(){
        $('<li>\
                <div class="input-group">\
                    <input type="text" class="form-control" placeholder="Parameter Key" data-rel="param-key">\
                    <span class="input-group-addon">=</span>\
                    <input type="text" class="form-control" placeholder="Parameter Value" data-rel="param-val">\
                </div>\
                <a href="javascript:void(0)" data-rel="param-remove"><span class="glyphicon glyphicon-remove-circle"></span></a>\
            </li>').appendTo($('#addParamList'))
    });

    $('#addParamList').delegate('a[data-rel="param-remove"]', 'click', function(){
        $(this).parent().remove();
    });



    $('#form').validate({
        rules: {
            message: {
                required: true
            },
            actiontype: {
                required: true
            }
        },
        errorElement: 'span',
        errorPlacement: function(error,el){
            el.parent().find('[data-control="form-input-error"]').append(error);
        },
        submitHandler:function(form) {
            var data = {
                message: form.message.value,
                user_ids: form.user_ids.value,
                openid_uins: form.openid_uins.value,
                push_env: $('#push_env').val(),
                actiontype: form.actiontype.value,
                actionparam: getParamList()
            };

            if(data.user_ids === "" && data.openid_uins === ""){
                IVYAlert("User ID or OpenID Uin is required!");
                return;
            }

            IVYAlert("Loading...", {
                toggle: true
            });

            // fix
            data.user_ids !== "" ? data.user_ids = JSON.stringify(data.user_ids.split('#')) : null;
            data.openid_uins !== "" ? data.openid_uins = JSON.stringify(data.openid_uins.split('#')) : null;

            $.post('/cms/index.php/msg_push/push', data, function(rs){
                if(rs.ret == 0){
                    IVYAlert(rs.msg);
                }else{
                    IVYAlert(rs.msg);
                }
            },'json');
        }
    });


    function getParamList(){
        var rs = [];
        var k, v;
        $('#addParamList').children().each(function(){
            var $this = $(this);
            k = $this.find('input[data-rel="param-key"]').val();
            v = $this.find('input[data-rel="param-val"]').val();
            if( k !== "" && v !== ""){
                var tmp = {};
                tmp[k] = v;
                rs.push(tmp);
            }
        });
        return (rs.length > 0 ? JSON.stringify(rs) : "");
    }
})
