// build-task/test/proj/resource/page/app/cuisine/js/cuisine.modify.js
/**
 * @author
 * @desc
 */
define(function(require, exports, module){
    var $ = require('jquery');
    var IVYAlert = require('ivy.alert');
    require('validation');

    $('#form').validate({
        rules: {
            en_name: {
                required: true
            },
            cn_name: {
                required: true
            }
        },
        errorElement: 'span',
        errorPlacement: function(error,el){
            el.parent().find('[data-control="form-input-error"]').append(error);
        },
        submitHandler:function(form) {
            IVYAlert("Loading...", {
                toggle: true
            });
            var data = {
                id: form.id.value,
                en_name: form.en_name.value,
                cn_name: form.cn_name.value,
                rank: form.rank.value,
                order: form.order.value,
                parent_id: $('#parent_id').val()
            };

            $.post('/cms/index.php/cuisine/domodify', data, function(rs){
                if(rs.ret == 0){
                    IVYAlert(rs.msg, function(){
                        location.href = "/cms/index.php/cuisine/index";
                    });
                }else{
                    IVYAlert(rs.msg);
                }
            },'json');
        }
    });
})
