// gruntjs/Gruntfile_.js

module.exports = function(grunt) {
    
	var style = require('grunt-cmd-transport').style.init(grunt);
	var jsParser = require('grunt-cmd-transport').script.init(grunt).jsParser;

    //===================================↓：配置区=========================================//
    var root = '__proj/mod';   // self define    // 静态配置构建任务
    grunt.initConfig({

        pkg: grunt.file.readJSON('package.json'),

        transport: {
            // eg: {
                // options: {
                    // idleading: '', 
                    // debug: false   
                // },
                // files: [{
                    // expand: true,
                    // cwd: root, 
                    // src: '**/*.js',
                    // dest: root + '_dist/'
                // }]
            // },
			// css2js: {
				// options: {
					// parsers: {
						// '.css': [style.css2jsParser]
					// },
					// debug: false
				// },
				// files: [{
				  // expand: true,
				  // cwd: root,
				  // src: '**/*.css',
				  // dest: root + '_dist/'
				// }]
			// },
			tpl: {
				files: [{
					expand: true,
					cwd: 'test',
					src: '**/*.tpl',
					dest: 'test_dist/'
				}]
			},
			
			// style: {
				// options: {
					// parsers: {
						// '.css': [style.css2jsParser],
						// '.js': [jsParser]
					// },
					// styleBox: [],
					// idleading: ''
				// },
				// files: [{
					// expand: true,
					// cwd: root,
					// src: '**/*.{js,css}',
					// dest: root + '_dist'
				// }]
			// },
			
			// css: {
				// options:{
					// debug: false
				// },
				// files: [{
					// expand: true,
					// cwd: root,
					// src: '**/*.css',
					// dest: root + '_dist
				// }]
			// },
			
			// style: {
				// options: {
					// parsers: {
						// '.css': [style.css2jsParser],
						// '.js': [jsParser]
					// },
					// styleBox: [],
					// idleading: '',
					// debug: false
				// },
				// files: [{
					// expand: true,
					// cwd: root,
					// src: '**/*.{js,css}',
					// dest: root + '_dist'
				// }]
			// },
        }
    });

    
    grunt.loadNpmTasks('grunt-cmd-transport');

    grunt.registerTask('default', ['transport']);

};
