// build-task/test/proj/resource/page/widget/foodie-preview/js/foodie-preview.js

define(function (require, exports, module) {

	var $ = require('jquery');
	require('bootstrap');
	require('../css/foodie-preview.css');

	$.fn.extend({

		foodiePreview: function (options) {

			options = options || {};

			this.each(function (index, element) {
				
				if (!$(this).data('foodiePreview')) {
					$(this).data('foodiePreview', {});
					
					var items = $(this).find('li').map(function (index, element) {
						return {
							title: $(this).attr('data-title'),
							iconUrl: $(this).attr('data-icon-url')
						};
					});
					
					$(this)
						.addClass('foodie-preview')
						.empty()
						.append($('<div class="preview"></div>')
							.append($('<ul class="clearfix"></ul>'))
						)
						.foodiePreview().addItem(items);
				}
			});

			return {

				$this: this,

				clear: function () {

					this.$this.find('.preview > ul')
						.empty();

					return this.$this;
				},

				addItem: function (item) {

					var container = this.$this;

					$(item).each(function (index, item) {
						container.find('.preview > ul')
							.append($('<li class="pull-left"></li>')
								.data('foodiePreview', {
									title: item.title,
									iconUrl: item.iconUrl
								})
								.append($('<img src="' + item.iconUrl + '" alt="' + item.title + '" />'))
								.append($('<p></p>')
									.html(item.title)
								)
							);
					});

					return this.$this;
				},

				getItemList: function () {

					return this.$this.find('.preview > ul > li').map(function (index, element) {
						return {
							title: $(this).data('foodiePreview').title,
							iconUrl: $(this).data('foodiePreview').iconUrl
						};
					}).get();
				}

			};
		}

	});

});
