// build-task/test/proj/resource/page/app/login/js/login.js

define(function (require, exports, module) {

	var $ = require('jquery');
	require('jquery.form');
	require('jquery-validate');
	require('bootstrap');
	var IVYAlert = require('ivy.alert');
	
	$('#signin-form').validate({
		errorPlacement: function (error, element) {
			if (element.attr('name') === 'login_name') ;
			if (element.attr('name') === 'password') ;
		}
	});

	$('#signin-form')
		.ajaxForm({
			beforeSerialize: function (data) {
				/*var urls = $(ue.getContent()).find('img').map(function (index, img) {
					return $(img).attr('src');
				}).get();
				$('#picture-urls').val(JSON.stringify(urls));*/
			},
			beforeSubmit: function (data) {
				IVYAlert('Signing in...', {
					toggle: true
				});
				/*$('#error-message').addClass('hidden');
				$('#success-message').addClass('hidden');*/
			},
			success: function (data) {
				data = $.parseJSON(data);
				if (data['ret']) {
					IVYAlert(data['msg']);
					// $('#error-message').html('Sorry, your submission failed. Please try again.');
					// $('#error-message').removeClass('hidden');
				} else {
					IVYAlert('Signed in Successfully', function () {
						window.location = '../index/home';
					});
					// $('#success-message').html('Your edit has been submitted.');
					// $('#success-message').removeClass('hidden');
				}
			}
		});

	$('#submit').removeAttr('disabled');

});
