// build-task/test/proj/resource/page/app/blogger_category/js/blogger.category.modify.js
/**
 * @author
 * @desc
 */
define(function(require, exports, module){
	var $ = require('jquery');
	var IVYAlert = require('ivy.alert');
	require('validation');
	require('bootstrap'); 
	
    // 初始化需要多语言添加的表单项
    // var IPICKFclangname = require('ipick.fc.langname');
    // var category = new IPICKFclangname();
    // category.init('#category', $('#category').children().children(),{
    //     add_btn_txt: 'Add Category',
    //     data: $('#category').attr('data-val') ? JSON.parse($('#category').attr('data-val')) : []
    // });
	
	var SWFUpload = require('swfupload');
	var swfConfig = require('page/app/search/js/search.modify.swf.config');

	// Icon
	new SWFUpload({
		// Backend Settings
		upload_url: "/cms/index.php/upload/up2tfs",
		post_params: {
			'user_name': 's_id'
		},

		// File Upload Settings
		file_size_limit : "2 MB",	// 2MB
		file_types : "*.jpg;*.png;*.gif",
		file_types_description : "JPG/PNG/GIF Images",
		file_upload_limit : "0",

		file_queue_error_handler : swfConfig.file_queue_error_handler,
		file_dialog_complete_handler : swfConfig.file_dialog_complete_handler,
		upload_progress_handler : swfConfig.upload_progress_handler, //
		upload_error_handler : swfConfig.upload_error_handler,
		upload_success_handler : function(file, rs){
			try{
				rs = JSON.parse(rs);
				if(rs.ret == 0){
					$('#category_icon').val(rs.data.pic_url);
					$('#upload_img').removeClass('hidden').attr('src', rs.data.pic_url);
					//网络慢的情况下，提示信息后很久才看到图片
					IVYAlert('Upload successfully!');
				}else{
					IVYAlert(rs.msg);
				}
			}catch(e){
				IVYAlert(rs);
			}
		},
		upload_complete_handler : swfConfig.upload_complete_handler,

		button_placeholder_id : "upload",
		button_width: 1000,
		button_height: 300,
		button_window_mode: SWFUpload.WINDOW_MODE.TRANSPARENT,
		button_cursor: SWFUpload.CURSOR.HAND,

		// Flash Settings
		flash_url : "/cms/resource/lib/cmd/swf/swfupload.swf",

		// Debug Settings
		debug: false
	});

	$('#form').validate({
		rules: {
			category_name:{
				required:true
			}
		},
		errorElement: 'span',
		errorPlacement: function(error,el){
		},
		submitHandler:function(form) {
			IVYAlert("Loading...", {
				toggle: true
			});
			var data = {
				id: form.id.value,
				name: form.category_name.value,
				icon_url: $('#category_icon').val(),
				description: form.description.value
			};
			$.post('/cms/index.php/article_category/do_modify', data, function(rs){
				if(rs.ret == 0){
					IVYAlert(rs.msg, function(){
						location.href = "/cms/index.php/article_category/index";
					});
				}else{
					IVYAlert(rs.msg);
				}
			},'json');
		}
	});
})
