// build-task/test/proj/resource/page/app/blogger_category/js/blogger.category.list.js
/**
 * @author
 * @desc
 */
define(function(require, exports, module){
    var $ = require('jquery');
    var IVYAlert = require('ivy.alert');
    var Pagingbar = require('pagingbar');

    require('ipick.foodie-preview');
    $('#modified-preview, #live-preview').foodiePreview();

    // Rewrite goto
    Pagingbar.gotoPage = function(pnum){
        if(pnum > 0 && pnum <= Pagingbar._data.total_page_num){
            Pagingbar.setPageNum(pnum);
            // Set index ui
            $('#common-pagingbar').find('li').removeClass('active');
            $('#common-pagingbar').find('a[data-val="' + pnum + '"]').parent().addClass('active');
            $.get('/cms/index.php/article_category/get_more', {
                page: Pagingbar.getPageNum(),
                num : Pagingbar.getPageSize()
            }, function(rs){
                $(window).scrollTop(0);
                var list = renderList(rs.data.data.list);
                if(list !== ""){
                    $('#homeList').children().remove();
                    $('#homeList').html(list);
                }
            },'json')
        }else{
            $('#common-pagingbar-goto').focus();
        }
    }
    // Init paging
    Pagingbar.init({
        page_size:  parseInt($('#common-pagingbar-page_size').val()),
        page_num:   parseInt($('#common-pagingbar-page_num').val()),
        total_page_num: parseInt($('#common-pagingbar-total_page_num').val()),
        country: $('#country').val()
    });
    /**
     * 刷新列表
     */
    var tplListData = require('page/app/blogger_category/js/blogger.category.list.data.js');
    function renderList(data){
        if(!data || data.length === 0){
            IVYAlert('No data');
            return "";
        }else{
            var html = "";
            var _idx = (Pagingbar.getPageNum() - 1) * Pagingbar.getPageSize() + 1;
            for(var i = 0, l = data.length;i < l;i++){
                html += tplListData.replace(/{index}/g, _idx + i)
                .replace(/{name}/g, data[i].name)
                .replace(/{icon_url}/g, data[i].icon_url?data[i].icon_url:'')
                .replace(/{status_cls}/g, data[i].status == 1 ? 'color_green' : 'color_red')
                .replace(/{status_val}/g, data[i].status == 1 ? 1 : 0)
                .replace(/{status_title}/g, data[i].status == 1 ? 'Click to offline' : 'Click to online')
                .replace(/{status_icon}/g, data[i].status == 1 ? 'glyphicon-ok-sign' : 'glyphicon-minus-sign"')
                .replace(/{last_update_h}/g, data[i].last_update_h)
                .replace(/{id}/g, data[i]._id_h)
                .replace(/{show_index}/g, data[i].show_index);
            }
            return html;
        }

        function disposeJSON(data){
            var rs = "";
            for(var itm in data){
                rs += '<div>' + itm + ': ' + (data[itm] ? data[itm] : '') + '</div>';
            }
            return rs;
        }
    }

    function formatPreviewData(online_list){
        var ret = '';
        if(online_list.length > 0){
            for(var i=0;i<online_list.length;i++){
                ret += '<li><img src="{icon_url}" /><span>{name}</span></li>'.replace(/\{(\w+)\}/ig,function(tag){
                    var tag_name = tag.substring(1,tag.length-1);
                    if(online_list[i][tag_name]){
                        return online_list[i][tag_name];
                    }else{
                        return '';
                    }
                });
            }
        }
        console.log(ret);
        return ret;
    }
    // Change Status
     $('#homeList').delegate('a[data-rel="status"]', 'click', function(){
        var $this = $(this);
        if($this.attr('data-val') == "0"){
            $.getJSON('/cms/index.php/article_category/online', {
                id: $this.attr('data-id')
            }, function(rs){
                if(rs.ret == 0){
                    // Change itself
                    $this.attr('data-val', 1).attr('title', 'Click to offline').removeClass('color_red').addClass('color_green').children().removeClass('glyphicon-minus-sign').addClass('glyphicon-ok-sign');
                    //                    $this.parent().parent().find('input[data-rel="sort"]').removeAttr('disabled');

                    // Render
                    $('#modified-preview').html(formatPreviewData(rs.data.online_list));
                }else{
                    IVYAlert(rs.msg);
                }
            })
        }else{
            $.getJSON('/cms/index.php/article_category/offline', {
                id: $this.attr('data-id')
            }, function(rs){
                if(rs.ret == 0){
                    // Change itself
                    $this.attr('data-val', 0).attr('title', 'Click to online').removeClass('color_green').addClass('color_red').children().removeClass('glyphicon-ok-sign').addClass('glyphicon-minus-sign');
                    //                    $this.parent().parent().find('input[data-rel="sort"]').attr('disabled', '');

                    // Render
                   $('#modified-preview').html(formatPreviewData(rs.data.online_list));
                }else{
                    IVYAlert(rs.msg);
                }
            })
        }
    });

    // Release
    $('#release-btn').click(function(){
        $.getJSON('/cms/index.php/release/article_category', "", function(rs){
            if(rs.ret == 0){
                IVYAlert(rs.msg, function(){
                    $('#modified-online').html($('#modified-preview').html());
                });
            }else{
                IVYAlert(rs.msg);
            }
        });
    })

    // Change show_index
    var SPage = {
        sortEvent: function(){
            $('#homeList').delegate('input[data-rel="sort"]', 'change', function(){
                var val = +parseInt(this.value);
                if(isNaN(val)){
                    return;
                }
                var $this = $(this);
                $.getJSON('/cms/index.php/article_category/modify_index', {
                    id: $this.attr('data-id'),
                    show_index: val
                }, function(rs){
                    $('#modified-preview').html(formatPreviewData(rs.data.online_list));
                })
            })
        }
    };
    SPage.sortEvent();
})
