// build-task/test/proj/resource/page/app/blogger_category/js/modify.js

define(function (require, exports, module) {

	var $ = require('jquery');
	require('jquery-validate');
	require('jquery.form');
	var IVYAlert = require('ivy.alert');

	require('ipick.file-upload');

	// Upload Icon
	$('#icon-upload').fileUpload({
		url: '/cms/index.php/upload/up2tfs',
		/*params: {
			'width': 50,
			'height': 50,
			'check_size': true
		},*/
		file: {
			types: '*.jpg;*.png;*.gif',
			typesDescription: 'JPG/PNG/GIF Images',
			sizeLimit: '2 MB',
		},
	}).on('uploadsuccess', function (event) {
		var data = $.parseJSON(event.body);
		if (data['ret']) {
			IVYAlert(data['msg']);
		} else {
			$('#icon').val(data['data']['pic_url']);
			$('#uploaded-icon').removeClass('hidden').attr('src', data['data']['pic_url']);
			IVYAlert('Icon uploaded successfully');
		}
	}).on('uploadstart', function (event) {
		IVYAlert('Uploading...', { toggle: true });
	}).on('error', function (event) {
		var message = {
			'QUEUE_ERROR.QUEUE_LIMIT_EXCEEDED': 'You can only upload one file at a time',
			'QUEUE_ERROR.ZERO_BYTE_FILE': 'The file is empty',
			'QUEUE_ERROR.FILE_EXCEEDS_SIZE_LIMIT': 'The file exceeded the maximum file size (2 MB)',
			'QUEUE_ERROR.INVALID_FILETYPE': 'The file format should be .bmp, .png, .jpeg, .jpg or .gif'
		}[event.error.type] || event.error.message;
		IVYAlert(message);
	});

	// Validation
	$('#form').validate({
		errorPlacement: function (error, element) {
		}
	});

	// Ajax Submit
	$('form')
		.ajaxForm({
			beforeSerialize: function (data) {
				/*if(!$('#icon').val()){
					IVYAlert('You should upload an icon before submit');
					return false;
				}*/
			},
			beforeSubmit: function (data) {
				IVYAlert('Submitting...', { toggle: true });
			},
			success: function (data) {
				data = $.parseJSON(data);
				if (data['ret']) {
					IVYAlert(data['msg']);
				} else {
					IVYAlert('Your edit has been submitted', function () {
						window.location = '../article_category/index';
					});
				}
			}
		});

});
